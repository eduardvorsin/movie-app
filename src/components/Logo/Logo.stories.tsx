import type { Meta, StoryObj } from '@storybook/react';
import Logo, { Props } from './Logo';
import { useTranslation } from 'react-i18next';

const sizes: NonNullable<Props['size']>[] = ['small', 'medium', 'large'];

const meta: Meta<typeof Logo> = {
	title: 'UI/Logo',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the Logo'
		},
		testId: {
			description: 'Id for testing the component',
		},
		size: {
			description: 'Logo Dimensions',
		},
		onClick: {
			control: false,
			description: 'A callback that is triggered when a component is clicked on',
		},
		dictionary: {
			control: false,
			description: 'An object with keys for localization of the component',
		},
	},
	component: Logo,
	parameters: {
		docs: {
			description: {
				component: 'The component responsible for displaying the project logo',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof Logo>;

const LogoWithHooks = (props: Omit<Props, 'dictionary'>) => {
	const { t } = useTranslation('common');
	const dictionary = {
		altText: t('logo.altText'),
		linkText: t('logo.linkText'),
	};

	return (
		<Logo
			{...props}
			dictionary={dictionary}
		/>
	);
}


export const Default: Story = {
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
	render: (args) => (<LogoWithHooks {...args} />),
};

export const EventCallbacks: Story = {
	argTypes: {
		onClick: {
			action: 'Clicked',
		},
	},
	parameters: {
		docs: {
			description: {
				story: 'When you click on the logo, a callback is triggered',
			},
		},
	},
	render: (args) => (<LogoWithHooks {...args} />),
};

export const Sizes: Story = {
	parameters: {
		docs: {
			description: {
				story: 'There are 3 logo sizes: small, medium, large.',
			},
		},
	},
	render: () => (
		<div
			className='flex flex-col gap-5'
		>
			{sizes.map((size) => (
				<LogoWithHooks
					key={size}
					size={size}
				/>
			))}
		</div>
	)
};
