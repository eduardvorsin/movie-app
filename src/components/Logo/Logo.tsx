'use client';

import Link from 'next/link';
import ThemedImage from '../ThemedImage/ThemedImage';
import { useTranslation } from '@/i18n/client';
import { useParams } from 'next/navigation';
import { Locales, fallbackLng } from '@/i18n/settings';
import { MouseEventHandler } from 'react';
import { routes } from 'src/constants';
import { GeneralProps } from '@/types/shared';

const sizes = {
	small: {
		width: 47,
		height: 24,
	},
	medium: {
		width: 62,
		height: 32,
	},
	large: {
		width: 110,
		height: 56,
	},
} as const;

export type Props = {
	onClick?: MouseEventHandler<HTMLAnchorElement>,
	size: 'small' | 'medium' | 'large',
} & GeneralProps;

export default function Logo({
	size = 'medium',
	className,
	testId,
	onClick,
	...props
}: Props) {
	const lang = useParams()?.lang as Locales ?? fallbackLng;
	const { t } = useTranslation(lang);

	return (
		<Link
			className={`inline-flex ${className}`}
			href={routes.home}
			data-testid={testId}
			onClick={onClick}
			{...props}
		>
			<ThemedImage
				width={sizes[size].width}
				height={sizes[size].height}
				alt='movie wander logo'
				src={{
					light: '/assets/images/logo.svg',
					dark: '/assets/images/logo.svg',
				}}
			/>
			<span className='sr-only'>
				{t('logoLink')}
			</span>
		</Link>
	);
};