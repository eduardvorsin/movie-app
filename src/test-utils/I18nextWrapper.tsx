import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;