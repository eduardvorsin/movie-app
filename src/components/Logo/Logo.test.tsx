import { render, screen } from "@testing-library/react";
import Logo, { Props } from "./Logo";
import i18next from "@/i18n/client";
import I18nextWrapper from "@/test-utils/I18nextWrapper";

const sizes: NonNullable<Props['size']>[] = ['small', 'medium', 'large'];

describe('Logo tests', () => {
	it('is rendered correctly', () => {
		const dictionary = {
			altText: 'MovieWander logo',
			linkText: 'To home page',
		};
		render(
			<Logo
				size='medium'
				dictionary={dictionary}
				testId='test-logo'
			/>
		);

		expect(screen.getByTestId<HTMLAnchorElement>('test-logo')).toBeInTheDocument();
	});

	it('is a basic snapshot', () => {
		const dictionary = {
			altText: 'MovieWander logo',
			linkText: 'To home page',
		};
		render(
			<Logo
				size='medium'
				dictionary={dictionary}
				testId='test-logo'
			/>
		);

		expect(screen.getByTestId<HTMLAnchorElement>('test-logo')).toMatchSnapshot();
	});

	it.each(sizes)('is a snapshot with a "%s" size', (size) => {
		const dictionary = {
			altText: 'MovieWander logo',
			linkText: 'To home page',
		};
		render(
			<Logo
				size={size}
				dictionary={dictionary}
				testId='test-logo'
			/>
		);

		expect(screen.getByTestId<HTMLAnchorElement>('test-logo')).toMatchSnapshot();
	});
});

describe('Banner integration tests', () => {
	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const translation = new RegExp(i18next.t('logo.linkText'));

		const dictionary = {
			altText: i18next.t('logo.altText'),
			linkText: i18next.t('logo.linkText'),
		};
		render(
			<Logo
				size='medium'
				dictionary={dictionary}
				testId='test-logo'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByText<HTMLSpanElement>(translation)).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const translation = new RegExp(i18next.t('logo.linkText'));

		const dictionary = {
			altText: i18next.t('logo.altText'),
			linkText: i18next.t('logo.linkText'),
		};
		render(
			<Logo
				size='medium'
				dictionary={dictionary}
				testId='test-logo'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByText<HTMLSpanElement>(translation)).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');

		const dictionary = {
			altText: i18next.t('logo.altText'),
			linkText: i18next.t('logo.linkText'),
		};
		render(
			<Logo
				size='medium'
				dictionary={dictionary}
				testId='test-logo'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByTestId<HTMLDivElement>('test-logo')).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');

		const dictionary = {
			altText: i18next.t('logo.altText'),
			linkText: i18next.t('logo.linkText'),
		};
		render(
			<Logo
				size='medium'
				dictionary={dictionary}
				testId='test-logo'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByTestId<HTMLDivElement>('test-logo')).toMatchSnapshot();
	});
});