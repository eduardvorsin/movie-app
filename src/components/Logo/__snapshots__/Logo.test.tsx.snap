// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logo integration tests is a snapshot with English localization 1`] = `
<a
  class="inline-flex undefined"
  data-testid="test-logo"
  href="/"
>
  <img
    alt="MovieWander logo"
    data-nimg="1"
    decoding="async"
    height="32"
    loading="lazy"
    src="http://localhost/assets/images/logo.svg"
    style="color: transparent;"
    width="62"
  />
  <span
    class="sr-only"
  >
    To home page
  </span>
</a>
`;

exports[`Logo integration tests is a snapshot with Russian localization 1`] = `
<a
  class="inline-flex undefined"
  data-testid="test-logo"
  href="/"
>
  <img
    alt="MovieWander логотип"
    data-nimg="1"
    decoding="async"
    height="32"
    loading="lazy"
    src="http://localhost/assets/images/logo.svg"
    style="color: transparent;"
    width="62"
  />
  <span
    class="sr-only"
  >
    На главную страницу
  </span>
</a>
`;

exports[`Logo tests is a basic snapshot 1`] = `
<a
  class="inline-flex undefined"
  data-testid="test-logo"
  href="/"
>
  <img
    alt="MovieWander logo"
    data-nimg="1"
    decoding="async"
    height="32"
    loading="lazy"
    src="http://localhost/assets/images/logo.svg"
    style="color: transparent;"
    width="62"
  />
  <span
    class="sr-only"
  >
    To home page
  </span>
</a>
`;

exports[`Logo tests is a snapshot with a "large" size 1`] = `
<a
  class="inline-flex undefined"
  data-testid="test-logo"
  href="/"
>
  <img
    alt="MovieWander logo"
    data-nimg="1"
    decoding="async"
    height="56"
    loading="lazy"
    src="http://localhost/assets/images/logo.svg"
    style="color: transparent;"
    width="110"
  />
  <span
    class="sr-only"
  >
    To home page
  </span>
</a>
`;

exports[`Logo tests is a snapshot with a "medium" size 1`] = `
<a
  class="inline-flex undefined"
  data-testid="test-logo"
  href="/"
>
  <img
    alt="MovieWander logo"
    data-nimg="1"
    decoding="async"
    height="32"
    loading="lazy"
    src="http://localhost/assets/images/logo.svg"
    style="color: transparent;"
    width="62"
  />
  <span
    class="sr-only"
  >
    To home page
  </span>
</a>
`;

exports[`Logo tests is a snapshot with a "small" size 1`] = `
<a
  class="inline-flex undefined"
  data-testid="test-logo"
  href="/"
>
  <img
    alt="MovieWander logo"
    data-nimg="1"
    decoding="async"
    height="24"
    loading="lazy"
    src="http://localhost/assets/images/logo.svg"
    style="color: transparent;"
    width="47"
  />
  <span
    class="sr-only"
  >
    To home page
  </span>
</a>
`;
